### PR TITLE
Get blocking time formula

### PR DESCRIPTION
**Pull request by 11michalis11**
- [x] Attempted to get recursive formula
- [x] Got formula by solving a system
- [x] Modularise
- [x] Test
- [x] Document

### CODE REVIEW

File: tex/main/MarkovChain/expressions_from_pi/main.tex [hunk 1]
```diff
@@ -328,3 +326,180 @@ \subsubsection{Overall Waiting Time}
...
+        c(u,v) + b(u, v-1), & \textbf{if } v = N \neq T \\
+        c(u,v) + p_s(u,v) b(u-1, v) + p_o(u,v) b(u, v+1), & \textbf{if } u > 0 \textbf{ and } v = T \\
+        c(u,v) + p_s(u,v) b(u, v-1) + p_o(u,v) b(u, v+1), & \textbf{otherwise} \\
+    \end{cases}
+\end{equation}
+
+Unlike equations (\ref{eq:waiting-time-at-each-state-other}) and 
+(\ref{eq:waiting-time-at-each-state-ambulance}), equation 
+(\ref{eq:blocking-time-at-each-state}) cannot be solved recursively. 
+An algebraic approach will be used to solve this equation here.
```
Comment (drvinceknight) [inline tex/main/MarkovChain/expressions_from_pi/main.tex:403]: Yeah algebraic isn't quite right. Arguably doing things recursively is algebraic.

I'm not entirely what the correct term is:

- Direct approach;
- Linear algebraic approach;

File: tex/main/MarkovChain/expressions_from_pi/main.tex [hunk 2]
```diff
@@ -328,3 +326,180 @@ \subsubsection{Overall Waiting Time}
...
+        c(u,v) + p_s(u,v) b(u-1, v) + p_o(u,v) b(u, v+1), & \textbf{if } u > 0 \textbf{ and } v = T \\
+        c(u,v) + p_s(u,v) b(u, v-1) + p_o(u,v) b(u, v+1), & \textbf{otherwise} \\
+    \end{cases}
+\end{equation}
+
+Unlike equations (\ref{eq:waiting-time-at-each-state-other}) and 
+(\ref{eq:waiting-time-at-each-state-ambulance}), equation 
+(\ref{eq:blocking-time-at-each-state}) cannot be solved recursively. 
+An algebraic approach will be used to solve this equation here. 
+By finding all equations \(b(u,v)\) for all states \((u,v)\) that belong in \(S_b\)
```
Comment (drvinceknight) [inline tex/main/MarkovChain/expressions_from_pi/main.tex:404]:
```suggestion
By enumerating all equations of (\ref{eq:blocking-time-at-each-state}) for all states \((u,v)\) that belong in \(S_b\) 
```

File: tex/main/MarkovChain/expressions_from_pi/main.tex [hunk 3]
```diff
@@ -328,3 +326,180 @@ \subsubsection{Overall Waiting Time}
...
+    \columnbreak
+    \vspace*{0cm} 
+    \begin{align*}
+        b(1,2) &= c(1,2) + p_o b(1,3) \\
+        b(1,3) &= c(1,3) + p_s b(1,2) + p_o b(1,4) \\
+        b(1,4) &= c(1,4) + b(1,3) \\
+        b(2,2) &= c(2,2) + p_s b(1,2) + p_o b(2,3) \\
+        b(2,3) &= c(2,3) + p_s b(2,2) + p_o b(1,4) \\
+        b(2,4) &= c(2,4) + b(2,3)
+    \end{align*}
```
Comment (drvinceknight) [inline tex/main/MarkovChain/expressions_from_pi/main.tex:423]:
```suggestion
    \begin{align}
        b(1,2) &= c(1,2) + p_o b(1,3) \label{eq:first_equation_of_example}
\\
        b(1,3) &= c(1,3) + p_s b(1,2) + p_o b(1,4) \\
        b(1,4) &= c(1,4) + b(1,3) \\
        b(2,2) &= c(2,2) + p_s b(1,2) + p_o b(2,3) \\
        b(2,3) &= c(2,3) + p_s b(2,2) + p_o b(1,4) \\
        b(2,4) &= c(2,4) + b(2,3)\label{eq:last_equation_of_example}
    \end{align}
```

File: tex/main/MarkovChain/expressions_from_pi/main.tex [hunk 4]
```diff
@@ -328,3 +326,180 @@ \subsubsection{Overall Waiting Time}
...
+By finding all equations \(b(u,v)\) for all states \((u,v)\) that belong in \(S_b\) 
+a system of linear equations arises where the unknown variables are all the \(b(u,v)\)
+terms.
+For instance, let us consider a Markov model where \(C=2, T=3, N=6, M=2\). 
+The Markov model and the equivalent equations are shown below. 
+The equations considered here are only the ones that correspond to the blocking 
+states.
+
+\begin{multicols*}{2}
+    \scalebox{0.68}{\input{MarkovChain/expressions_from_pi/example_model_2362/main.tex}}
```
Comment (drvinceknight) [inline tex/main/MarkovChain/expressions_from_pi/main.tex:413]: Turn this in to a figure environment.

File: tex/main/MarkovChain/expressions_from_pi/main.tex [hunk 5]
```diff
@@ -328,3 +326,180 @@ \subsubsection{Overall Waiting Time}
...
+
+Unlike equations (\ref{eq:waiting-time-at-each-state-other}) and 
+(\ref{eq:waiting-time-at-each-state-ambulance}), equation 
+(\ref{eq:blocking-time-at-each-state}) cannot be solved recursively. 
+An algebraic approach will be used to solve this equation here. 
+By finding all equations \(b(u,v)\) for all states \((u,v)\) that belong in \(S_b\) 
+a system of linear equations arises where the unknown variables are all the \(b(u,v)\)
+terms.
+For instance, let us consider a Markov model where \(C=2, T=3, N=6, M=2\). 
+The Markov model and the equivalent equations are shown below.
```
Comment (drvinceknight) [inline tex/main/MarkovChain/expressions_from_pi/main.tex:408]:
```suggestion
The Markov model is shown in Figure ... and the equivalent equations are (\ref{eq:first_equation_of_example})-(eq:last_equation_of_example).
```

Avoid referring to things based on their position. Their position can change.

File: tex/main/MarkovChain/expressions_from_pi/main.tex [hunk 6]
```diff
@@ -328,3 +326,180 @@ \subsubsection{Overall Waiting Time}
...
+        b(1,2) &= c(1,2) + p_o b(1,3) \\
+        b(1,3) &= c(1,3) + p_s b(1,2) + p_o b(1,4) \\
+        b(1,4) &= c(1,4) + b(1,3) \\
+        b(2,2) &= c(2,2) + p_s b(1,2) + p_o b(2,3) \\
+        b(2,3) &= c(2,3) + p_s b(2,2) + p_o b(1,4) \\
+        b(2,4) &= c(2,4) + b(2,3)
+    \end{align*}
+\end{multicols*}
+
+Additionally, the above equations can be transformed into the following linear system.
```
Comment (drvinceknight) [inline tex/main/MarkovChain/expressions_from_pi/main.tex:426]:
```suggestion
Additionally, the above equations can be transformed into a linear system of the form \(Mx=y\) with:
```

File: tex/main/MarkovChain/expressions_from_pi/main.tex [hunk 7]
```diff
@@ -328,3 +326,180 @@ \subsubsection{Overall Waiting Time}
...
+
+\begin{equation}\label{eq:example-algebaric-approach-blocking-time}
+    \begin{pmatrix}
```
Comment (drvinceknight) [inline tex/main/MarkovChain/expressions_from_pi/main.tex:429]:
```suggestion
    M=
    \begin{pmatrix}
```

File: tex/main/MarkovChain/expressions_from_pi/main.tex [hunk 8]
```diff
@@ -328,3 +326,180 @@ \subsubsection{Overall Waiting Time}
...
+        -1 & p_o & 0 & 0 & 0 & 0 \\ %(1,2)
+        p_s & -1 & p_o & 0 & 0 & 0 \\ %(1,3)
+        0 & 1 & -1 & 0 & 0 & 0 \\ %(1,4)
+        p_s & 0 & 0 & -1 & p_o & 0\\ %(2,2)
+        0 & 0 & 0 & p_s & -1 & p_o \\ %(2,3)
+        0 & 0 & 0 & 0 & 1 & -1 \\ %(2,4)
+    \end{pmatrix}
+    \begin{pmatrix}
```
Comment (drvinceknight) [inline tex/main/MarkovChain/expressions_from_pi/main.tex:437]:
```suggestion
    x=
    \begin{pmatrix}
```

File: tex/main/MarkovChain/expressions_from_pi/main.tex [hunk 9]
```diff
@@ -328,3 +326,180 @@ \subsubsection{Overall Waiting Time}
...
+        b(1,2) \\
+        b(1,3) \\
+        b(1,4) \\
+        b(2,2) \\
+        b(2,3) \\
+        b(2,4) \\
+    \end{pmatrix}
+    = 
+    \begin{pmatrix}
```
Comment (drvinceknight) [inline tex/main/MarkovChain/expressions_from_pi/main.tex:446]:
```suggestion
    y=
    \begin{pmatrix}
```

File: tex/main/MarkovChain/expressions_from_pi/main.tex (hunk 6 above)
Comment (drvinceknight) [reply tex/main/MarkovChain/expressions_from_pi/main.tex:426]: Not sure if `M, x, y` are the best ones but something like this.

File: tex/main/MarkovChain/expressions_from_pi/main.tex [hunk 10]
```diff
@@ -328,3 +326,180 @@ \subsubsection{Overall Waiting Time}
...
+        -c(1,3) \\
+        -c(1,4) \\
+        -c(2,2) \\
+        -c(2,3) \\
+        -c(2,4) \\
+    \end{pmatrix}
+\end{equation}
+
+A more generalised form of equation \ref{eq:example-algebaric-approach-blocking-time}
+can thus be given for any value of \(C,T,N,M\) by:
```
Comment (drvinceknight) [inline tex/main/MarkovChain/expressions_from_pi/main.tex:457]: Yeah, `M` is for sure the wrong letter for my suggestion for the linear system.

File: tex/main/MarkovChain/expressions_from_pi/main.tex [hunk 11]
```diff
@@ -328,3 +326,180 @@ \subsubsection{Overall Waiting Time}
...
+
+\begin{equation}\label{eq:general-algebaric-approach-blocking-time}
```
Comment (drvinceknight) [inline tex/main/MarkovChain/expressions_from_pi/main.tex:459]: So before this write out explicitly/rigorously the terms of the linear system and then go: `which will resemble:` followed by the `...` stuff.


(Let me know if that's not clear)

**Review by drvinceknight — CHANGES_REQUESTED** [review]
This is well written @11michalis11, some requests from me for the writing.

I've also suggested places where we should for now add `TODO` to keep track (I might have missed others but good to do this) and going forward only write using whatever vocabulary we plan to keep.

File: tex/main/MarkovChain/expressions_from_pi/main.tex [hunk 12]
```diff
@@ -328,3 +326,180 @@ \subsubsection{Overall Waiting Time}
...
+
+\begin{equation}\label{eq:blocking-time-at-each-state}
+    b(u,v) = 
+    \begin{cases} 
+        0, & \textbf{if } (u,v) \notin S_b \\
+        c(u,v) + b(u - 1, v), & \textbf{if } v = N = T\\
+        c(u,v) + b(u, v-1), & \textbf{if } v = N \neq T \\
+        c(u,v) + p_s(u,v) b(u-1, v) + p_o(u,v) b(u, v+1), & \textbf{if } u > 0 \textbf{ and } v = T \\
+        c(u,v) + p_s(u,v) b(u, v-1) + p_o(u,v) b(u, v+1), & \textbf{otherwise} \\
+    \end{cases}
```
[the remark this answers is not in the document]
Comment (11michalis11) [reply tex/main/MarkovChain/expressions_from_pi/main.tex:400]: Cool. So a Markov model that shows exactly how these expressions are formed?

File: tex/main/MarkovChain/expressions_from_pi/main.tex (hunk 1 above)
Comment (11michalis11) [reply tex/main/MarkovChain/expressions_from_pi/main.tex:406]: I agree, algebraic may not be the right name for it. Should we settle for linear algebraic approach for now? (I think I like it slightly better). There may even be a third name that describes it a better though.

File: tex/main/MarkovChain/expressions_from_pi/main.tex (hunk 10 above)
Comment (11michalis11) [reply tex/main/MarkovChain/expressions_from_pi/main.tex:464]: I'm gonna use `Z`. Running out of letters though 😆

File: tex/main/MarkovChain/expressions_from_pi/main.tex (hunk 11 above)
Comment (11michalis11) [reply tex/main/MarkovChain/expressions_from_pi/main.tex:466]: So write down the actual generalised equations before the matrix? (similar to what I did with the example model)

File: tex/main/MarkovChain/expressions_from_pi/main.tex (hunk 1 above)
Comment (11michalis11) [reply tex/main/MarkovChain/expressions_from_pi/main.tex:406]: Could even be `Solving the linear system approach`

File: tex/main/MarkovChain/expressions_from_pi/main.tex (hunk 10 above)
Comment (drvinceknight) [reply tex/main/MarkovChain/expressions_from_pi/main.tex:464]: `Z` is not idea (it's "calculusy") but yeah go with that, we can always change at a later date :) :+1:

File: tex/main/MarkovChain/expressions_from_pi/main.tex (hunk 11 above)
Comment (drvinceknight) [reply tex/main/MarkovChain/expressions_from_pi/main.tex:466]: Yup :+1:

**Pull request merged**